File: data_structures/doubly_linked_list/python/test_doubly_linked_list.py
import pytest
from doubly_linked_list import DoublyLinkedList


def test_new_is_empty():
    ll = DoublyLinkedList()
    assert ll.is_empty()
    assert ll.size() == 0


def test_push_front_order():
    ll = DoublyLinkedList()
    ll.push_front(3)
    ll.push_front(2)
    ll.push_front(1)
    assert ll.to_list() == [1, 2, 3]


def test_push_back_order():
    ll = DoublyLinkedList()
    ll.push_back(1)
    ll.push_back(2)
    ll.push_back(3)
    assert ll.to_list() == [1, 2, 3]


def test_front_back():
    ll = DoublyLinkedList()
    ll.push_back(1)
    ll.push_back(2)
    ll.push_back(3)
    assert ll.front() == 1
    assert ll.back() == 3


def test_pop_front_and_back():
    ll = DoublyLinkedList()
    ll.push_back(1)
    ll.push_back(2)
    ll.push_back(3)
    assert ll.pop_front() == 1
    assert ll.front() == 2
    assert ll.pop_back() == 3
    assert ll.back() == 2
    assert ll.size() == 1


def test_contains():
    ll = DoublyLinkedList()
    ll.push_back(1)
    ll.push_back(2)
    ll.push_back(3)
    assert ll.contains(2)
    assert not ll.contains(5)


def test_remove():
    ll = DoublyLinkedList()
    ll.push_back(1)
    ll.push_back(2)
    ll.push_back(3)
    assert ll.remove(2)
    assert ll.to_list() == [1, 3]
    assert ll.remove(1)
    assert ll.to_list() == [3]
    assert not ll.remove(99)


def test_front_empty_raises():
    with pytest.raises(IndexError):
        DoublyLinkedList().front()


def test_pop_back_empty_raises():
    with pytest.raises(IndexError):
        DoublyLinkedList().pop_back()

File: data_structures/doubly_linked_list/matlab/test_doubly_linked_list.m
% test_doubly_linked_list.m

ll = doubly_linked_list('new');
assert(doubly_linked_list(ll, 'is_empty'));

ll = doubly_linked_list(ll, 'push_front', 3);
ll = doubly_linked_list(ll, 'push_front', 2);
ll = doubly_linked_list(ll, 'push_front', 1);
assert(isequal(doubly_linked_list(ll, 'to_array'), [1, 2, 3]));

ll = doubly_linked_list('new');
ll = doubly_linked_list(ll, 'push_back', 1);
ll = doubly_linked_list(ll, 'push_back', 2);
ll = doubly_linked_list(ll, 'push_back', 3);
assert(isequal(doubly_linked_list(ll, 'to_array'), [1, 2, 3]));
assert(doubly_linked_list(ll, 'front') == 1);
assert(doubly_linked_list(ll, 'back')  == 3);

[ll, v] = doubly_linked_list(ll, 'pop_front');
assert(v == 1);
[ll, v] = doubly_linked_list(ll, 'pop_back');
assert(v == 3);
assert(doubly_linked_list(ll, 'size') == 1);

ll = doubly_linked_list('new');
ll = doubly_linked_list(ll, 'push_back', 1);
ll = doubly_linked_list(ll, 'push_back', 2);
ll = doubly_linked_list(ll, 'push_back', 3);
assert(doubly_linked_list(ll, 'contains', 2));
assert(~doubly_linked_list(ll, 'contains', 5));

[ll, ok] = doubly_linked_list(ll, 'remove', 2);
assert(ok);
assert(isequal(doubly_linked_list(ll, 'to_array'), [1, 3]));
[ll, ok] = doubly_linked_list(ll, 'remove', 99);
assert(~ok);

ll_empty = doubly_linked_list('new');
try
    doubly_linked_list(ll_empty, 'front');
    error('expected error not thrown');
catch e
    assert(strcmp(e.identifier, 'doubly_linked_list:underflow'));
end

disp('All tests passed.');
